tools/monitoring/config_test.go

```diff
@@ -0,0 +1,228 @@
+/*
+Copyright 2019 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package main
+
+import (
+	"reflect"
+	"testing"
+)
+
+const (
+	sampleYaml = `spec:
+  - error-pattern: 'Something went wrong: starting e2e cluster: error creating cluster'
+    hint: 'Check gcp status'
+    alerts:
+      - job-name-regex: '^pull.*'
+        occurrences: 2
+        jobs-affected: 1
+        prs-affected: 2
+        period: 1440 # 1440 minutes = 24 hours
+      - job-name-regex: '.*'
+        occurrences: 5
+        jobs-affected: 2
+        prs-affected: 1 # for non-pull jobs, we don't care about the number of prs affected, so we set the number to 1, which will basically make this particular condition always true
+        period: 1440
+
+  - error-pattern: 'sample*error2'
+    hint: 'hint_for_pattern_2'
+    alerts:
+      - job-name-regex: '^pull.*'
```

Comment:
What happens if there's a missing field e.g. jobs-affected? What is the error propagated? Or is it ignoring that particular error pattern if that's the case?

Reply:
We should have reasonable defaults for all the fields that are not madatory and check the ones that need a value

Reply:
I agree we can add defaults for some fields. Yet we should do that in the source code, and we should add a github task for that to do later. This pr is for unit test and we shouldn't add the check here.

Reply:
@yt3liu some updates after our offline conversation. I check with both Unmarshal and UnmarshalStrict. None of them catches missing fields as errors. Instead, they use zero values for the type as defaults. Thus, if we want to have our own defaults and/or check missing fields, we should add a validator/processor on the returned structure of newConfig.

That said, UnmarshalStrict will report error on any field name that does not appear in the unmarshall target struct. So it will help us find some errors in the config such as field misspelling.

Reply:
Thanks for looking into this. 

In this case, I think we do need to have an issue for setting the reasonable defaults so we don't accidentally set the state to always alarm (occurrence, prs-affected, jobs-affected=0) or never alarm (period=0). I think we can merge this in after we have an issue for to track it. 

I can't think of a use case for when the config.yaml would have fields not in the struct and still be a valid one except for the case where we'd like to update the names of the key i.e. job-name-regex to job-name-match-string, then config.yaml can be temporarily invalid if we use UnmarshalStrict. But I don't see we'll want to do that in the future, so do you think it make sense to change `Unmarshal` to `UnmarshalStrict`? 
(You can do that in a separate PR if you agree UnmarshalStrict handles better error catching than Unmarshal) 

Please create an issue to set the reasonable defaults referencing this issue.

Reply:
tracking issue: https://github.com/knative/test-infra/issues/810

Yes, I agree UnmarshalStrict is better